File: src/complier/contract/compiler.py
"""Compiler from parsed contract specs to runtime contract objects."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from .ast import (
    BranchStep,
    ElseArm,
    ForkStep,
    Guarantee,
    HumanStep,
    JoinStep,
    LlmStep,
    LoopStep,
    Param,
    ParamValue,
    Program,
    ProseGuard,
    Step,
    SubworkflowStep,
    ToolStep,
    UnorderedCase,
    UnorderedStep,
    WhenArm,
    Workflow,
)
from .model import Contract
from .parser import ParsedContract
from .runtime import (
    BranchBackNode,
    BranchNode,
    CallNode,
    CompiledWorkflow,
    EndNode,
    ForkNode,
    HumanNode,
    JoinNode,
    LLMNode,
    StartNode,
    ToolNode,
    UnorderedBackNode,
    UnorderedNode,
)


@dataclass(slots=True)
class CompileResult:
    """Entry and exit points for a compiled step sequence."""

    entry_id: str
    exit_ids: list[str]


@dataclass(slots=True)
class WorkflowCompiler:
    """Compiles workflow AST nodes into runtime graphs."""

    guarantees: dict[str, ProseGuard]
    workflow_name: str
    nodes: dict[str, object] = field(default_factory=dict)
    counter: int = 0

    def compile_workflow(self, workflow: Workflow) -> CompiledWorkflow:
        inherited_guards = [self._resolve_guarantee(name) for name in workflow.always]

        start = self._add_node(StartNode(id=self._new_id("start")))
        end = self._add_node(EndNode(id=self._new_id("end")))

        if workflow.steps:
            compiled = self._compile_steps(workflow.steps, inherited_guards)
            start.next_ids.append(compiled.entry_id)
            for exit_id in compiled.exit_ids:
                self.nodes[exit_id].next_ids.append(end.id)
        else:
            start.next_ids.append(end.id)

        return CompiledWorkflow(
            name=workflow.name,
            start_node_id=start.id,
            end_node_id=end.id,
            nodes=self.nodes,
        )

    def _compile_steps(
        self, steps: list[Step], inherited_guards: list[ProseGuard]
    ) -> CompileResult:
        compiled_steps = [self._compile_step(step, inherited_guards) for step in steps]

        entry_id = compiled_steps[0].entry_id
        pending = deque(compiled_steps[0].exit_ids)
        for compiled in compiled_steps[1:]:
            while pending:
                self.nodes[pending.popleft()].next_ids.append(compiled.entry_id)
            pending.extend(compiled.exit_ids)

        return CompileResult(entry_id=entry_id, exit_ids=list(pending))

    def _compile_step(
        self, step: Step, inherited_guards: list[ProseGuard]
    ) -> CompileResult:
        if isinstance(step, ToolStep):
            node = self._add_node(
                ToolNode(
                    id=self._new_id("tool"),
                    tool_name=step.name,
                    params={param.name: self._resolve_param_value(param.value) for param in step.params},
                    guards=list(inherited_guards),
                )
            )
            return CompileResult(entry_id=node.id, exit_ids=[node.id])

        if isinstance(step, HumanStep):
            node = self._add_node(
                HumanNode(
                    id=self._new_id("human"),
                    prompt=step.prompt,
                    guards=list(inherited_guards),
                )
            )
            return CompileResult(entry_id=node.id, exit_ids=[node.id])

        if isinstance(step, LlmStep):
            node = self._add_node(
                LLMNode(
                    id=self._new_id("llm"),
                    prompt=step.prompt,
                    guards=list(inherited_guards),
                )
            )
            return CompileResult(entry_id=node.id, exit_ids=[node.id])

        if isinstance(step, SubworkflowStep):
            node = self._add_node(
                CallNode(
                    id=self._new_id("call"),
                    call_type=step.call_type,
                    workflow_name=step.workflow_name,
                    guards=list(inherited_guards),
                )
            )
            return CompileResult(entry_id=node.id, exit_ids=[node.id])

        if isinstance(step, ForkStep):
            node = self._add_node(
                ForkNode(
                    id=self._new_id("fork"),
                    fork_id=step.fork_id,
                    call_type=step.target.call_type,
                    workflow_name=step.target.workflow_name,
                    guards=list(inherited_guards),
                )
            )
            return CompileResult(entry_id=node.id, exit_ids=[node.id])

        if isinstance(step, JoinStep):
            node = self._add_node(
                JoinNode(
                    id=self._new_id("join"),
                    fork_id=step.fork_id,
                    guards=list(inherited_guards),
                )
            )
            return CompileResult(entry_id=node.id, exit_ids=[node.id])

        if isinstance(step, BranchStep):
            return self._compile_branch(step, inherited_guards)

        if isinstance(step, LoopStep):
            return self._compile_loop(step, inherited_guards)

        if isinstance(step, UnorderedStep):
            return self._compile_unordered(step, inherited_guards)

        raise TypeError(f"Unsupported step type: {type(step)!r}")

    def _compile_branch(
        self, step: BranchStep, inherited_guards: list[ProseGuard]
    ) -> CompileResult:
        back = self._add_node(BranchBackNode(id=self._new_id("branch_back")))
        branch = self._add_node(
            BranchNode(
                id=self._new_id("branch"),
                branch_back_id=back.id,
            )
        )

        for arm in step.when_arms:
            compiled = self._compile_steps(arm.steps, inherited_guards)
            branch.arms[arm.condition] = compiled.entry_id
            for exit_id in compiled.exit_ids:
                self.nodes[exit_id].next_ids.append(back.id)

        if step.else_arm is not None:
            compiled_else = self._compile_steps(step.else_arm.steps, inherited_guards)
            branch.else_node_id = compiled_else.entry_id
            for exit_id in compiled_else.exit_ids:
                self.nodes[exit_id].next_ids.append(back.id)
        else:
            branch.else_node_id = back.id

        return CompileResult(entry_id=branch.id, exit_ids=[back.id])

    def _compile_loop(
        self, step: LoopStep, inherited_guards: list[ProseGuard]
    ) -> CompileResult:
        back = self._add_node(BranchBackNode(id=self._new_id("loop_back")))
        branch = self._add_node(
            BranchNode(
                id=self._new_id("loop_branch"),
                mode="loop",
                loop_until=step.until,
                branch_back_id=back.id,
            )
        )

        compiled_body = self._compile_steps(step.steps, inherited_guards)
        branch.arms[step.until] = back.id
        branch.else_node_id = compiled_body.entry_id

        for exit_id in compiled_body.exit_ids:
            self.nodes[exit_id].next_ids.append(branch.id)

        return CompileResult(entry_id=branch.id, exit_ids=[back.id])

    def _compile_unordered(
        self, step: UnorderedStep, inherited_guards: list[ProseGuard]
    ) -> CompileResult:
        back = self._add_node(UnorderedBackNode(id=self._new_id("unordered_back")))
        unordered = self._add_node(
            UnorderedNode(
                id=self._new_id("unordered"),
                back_node_id=back.id,
            )
        )

        for case in step.cases:
            compiled = self._compile_steps(case.steps, inherited_guards)
            unordered.case_entry_ids[case.label] = compiled.entry_id
            for exit_id in compiled.exit_ids:
                self.nodes[exit_id].next_ids.append(back.id)

        return CompileResult(entry_id=unordered.id, exit_ids=[back.id])

    def _resolve_param_value(self, value: ParamValue) -> ParamValue:
        if isinstance(value, (str, int, bool)) or value is None:
            return value
        return value  # ProseGuard — no resolution needed

    def _resolve_guarantee(self, name: str) -> ProseGuard:
        if name not in self.guarantees:
            raise ValueError(f"Unknown guarantee reference: {name}")
        return self.guarantees[name]

    def _new_id(self, prefix: str) -> str:
        self.counter += 1
        return f"{self.workflow_name}:{prefix}:{self.counter}"

    def _add_node(self, node: object) -> object:
        self.nodes[node.id] = node
        return node


@dataclass(slots=True)
class ContractCompiler:
    """Compiles parsed contract data into a runtime-ready contract."""

    def compile(self, parsed: ParsedContract) -> Contract:
        """Build the runtime contract from a parsed representation."""
        if not isinstance(parsed, ParsedContract):
            raise TypeError("Parsed contract must be a ParsedContract instance.")

        program = parsed.program
        if not isinstance(program, Program):
            raise TypeError("Parsed contract must contain a Program AST.")

        guarantees = {
            item.name: item.expression
            for item in program.items
            if isinstance(item, Guarantee)
        }
        workflows = [
            item
            for item in program.items
            if isinstance(item, Workflow)
        ]

        compiled_workflows = {
            workflow.name: WorkflowCompiler(
                guarantees=guarantees,
                workflow_name=workflow.name,
            ).compile_workflow(workflow)
            for workflow in workflows
        }

        return Contract(
            name="anonymous",
            workflows=compiled_workflows,
            guarantees=guarantees,
            metadata={
                "source": parsed.source,
                "parse_tree": parsed.tree,
                "program": program,
            },
        )

File: csm/src/compiler.cpp
#include "complier/compiler.hpp"

#include <format>
#include <utility>

#include "complier/error.hpp"
#include "complier/runtime.hpp"

namespace complier {

namespace {

// entry and exit points of a compiled step sequence
struct Wiring {
    std::string entry_id;
    std::vector<std::string> exit_ids;
};

struct WorkflowCompiler {
    const std::map<std::string, ast::Constraint>& guarantees;
    const std::string& workflow_name;
    std::map<std::string, rt::Node> nodes;
    int counter = 0;

    std::string new_id(const char* prefix) {
        return std::format("{}:{}:{}", workflow_name, prefix, ++counter);
    }

    rt::Node& add(const char* prefix, rt::NodeData data,
                  std::vector<ast::Constraint> guards = {}) {
        std::string id = new_id(prefix);
        auto [it, inserted] = nodes.emplace(
            id, rt::Node{id, {}, std::move(guards), std::move(data)});
        return it->second;
    }

    const ast::Constraint& resolve_guarantee(const std::string& name) const {
        auto it = guarantees.find(name);
        if (it == guarantees.end())
            throw CompileError(std::format("unknown guarantee reference: {}", name));
        return it->second;
    }

    rt::CompiledWorkflow compile(const ast::Workflow& workflow) {
        std::vector<ast::Constraint> guards;
        for (const auto& name : workflow.always) guards.push_back(resolve_guarantee(name));

        rt::Node& start = add("start", rt::StartNode{});
        std::string end_id = add("end", rt::EndNode{}).id;

        if (!workflow.steps.empty()) {
            Wiring wiring = compile_steps(workflow.steps, guards);
            start.next_ids.push_back(wiring.entry_id);
            for (const auto& exit_id : wiring.exit_ids)
                nodes.at(exit_id).next_ids.push_back(end_id);
        } else {
            start.next_ids.push_back(end_id);
        }

        rt::CompiledWorkflow out;
        out.name = workflow.name;
        out.start_node_id = start.id;
        out.end_node_id = std::move(end_id);
        out.ambient.insert(workflow.ambient.begin(), workflow.ambient.end());
        out.nodes = std::move(nodes);
        return out;
    }

    Wiring compile_steps(const std::vector<ast::Step>& steps,
                         const std::vector<ast::Constraint>& guards) {
        Wiring first = compile_step(steps[0], guards);
        std::vector<std::string> pending = std::move(first.exit_ids);
        for (std::size_t i = 1; i < steps.size(); ++i) {
            Wiring next = compile_step(steps[i], guards);
            for (const auto& exit_id : pending)
                nodes.at(exit_id).next_ids.push_back(next.entry_id);
            pending = std::move(next.exit_ids);
        }
        return {first.entry_id, std::move(pending)};
    }

    Wiring compile_step(const ast::Step& step, const std::vector<ast::Constraint>& guards) {
        if (const auto* tool = std::get_if<ast::ToolStep>(&step.value)) {
            std::map<std::string, ast::ParamValue> params;
            for (const auto& param : tool->params) params[param.name] = param.value;
            rt::Node& node = add("tool", rt::ToolNode{tool->name, std::move(params)}, guards);
            return {node.id, {node.id}};
        }
        if (const auto* human = std::get_if<ast::HumanStep>(&step.value)) {
            rt::Node& node = add("human", rt::HumanNode{human->prompt}, guards);
            return {node.id, {node.id}};
        }
        if (const auto* llm = std::get_if<ast::LlmStep>(&step.value)) {
            rt::Node& node = add("llm", rt::LlmNode{llm->prompt}, guards);
            return {node.id, {node.id}};
        }
        if (const auto* call = std::get_if<ast::SubworkflowStep>(&step.value)) {
            rt::Node& node =
                add("call", rt::CallNode{call->call_type, call->workflow_name}, guards);
            return {node.id, {node.id}};
        }
        if (const auto* fork = std::get_if<ast::ForkStep>(&step.value)) {
            rt::Node& node = add(
                "fork",
                rt::ForkNode{fork->fork_id, fork->target.call_type, fork->target.workflow_name},
                guards);
            return {node.id, {node.id}};
        }
        if (const auto* join = std::get_if<ast::JoinStep>(&step.value)) {
            rt::Node& node = add("join", rt::JoinNode{join->fork_id}, guards);
            return {node.id, {node.id}};
        }
        if (const auto* branch = std::get_if<ast::BranchStep>(&step.value))
            return compile_branch(*branch, guards);
        if (const auto* loop = std::get_if<ast::LoopStep>(&step.value))
            return compile_loop(*loop, guards);
        const auto& unordered = std::get<ast::UnorderedStep>(step.value);
        return compile_unordered(unordered, guards);
    }

    Wiring compile_branch(const ast::BranchStep& step,
                          const std::vector<ast::Constraint>& guards) {
        rt::Node& back = add("branch_back", rt::BranchBackNode{});
        rt::Node& branch = add("branch", rt::BranchNode{});
        auto& data = std::get<rt::BranchNode>(branch.data);
        data.branch_back_id = back.id;

        for (const auto& arm : step.when_arms) {
            Wiring wiring = compile_steps(arm.steps, guards);
            data.arms.emplace_back(arm.condition, wiring.entry_id);
            for (const auto& exit_id : wiring.exit_ids)
                nodes.at(exit_id).next_ids.push_back(back.id);
        }

        if (step.else_arm) {
            Wiring wiring = compile_steps(step.else_arm->steps, guards);
            data.else_node_id = wiring.entry_id;
            for (const auto& exit_id : wiring.exit_ids)
                nodes.at(exit_id).next_ids.push_back(back.id);
        } else {
            data.else_node_id = back.id;
        }

        return {branch.id, {back.id}};
    }

    // a loop is a branch whose until-arm jumps straight to the back node and
    // whose else-arm re-enters the body
    Wiring compile_loop(const ast::LoopStep& step, const std::vector<ast::Constraint>& guards) {
        rt::Node& back = add("loop_back", rt::BranchBackNode{});
        rt::Node& branch = add("loop_branch", rt::BranchNode{});
        auto& data = std::get<rt::BranchNode>(branch.data);
        data.is_loop = true;
        data.loop_until = step.until;
        data.branch_back_id = back.id;

        Wiring body = compile_steps(step.steps, guards);
        data.arms.emplace_back(step.until, back.id);
        data.else_node_id = body.entry_id;
        for (const auto& exit_id : body.exit_ids)
            nodes.at(exit_id).next_ids.push_back(branch.id);

        return {branch.id, {back.id}};
    }

    Wiring compile_unordered(const ast::UnorderedStep& step,
                             const std::vector<ast::Constraint>& guards) {
        rt::Node& back = add("unordered_back", rt::UnorderedBackNode{});
        rt::Node& unordered = add("unordered", rt::UnorderedNode{});
        auto& data = std::get<rt::UnorderedNode>(unordered.data);
        data.back_node_id = back.id;

        for (const auto& c : step.cases) {
            Wiring wiring = compile_steps(c.steps, guards);
            data.cases.emplace_back(c.label, wiring.entry_id);
            for (const auto& exit_id : wiring.exit_ids)
                nodes.at(exit_id).next_ids.push_back(back.id);
        }

        return {unordered.id, {back.id}};
    }
};

}  // namespace

Contract compile(const ast::Program& program, std::string_view source) {
    std::map<std::string, ast::Constraint> guarantees;
    for (const auto& item : program.items)
        if (const auto* g = std::get_if<ast::Guarantee>(&item)) guarantees[g->name] = g->expression;

    Contract contract;
    contract.name = "anonymous";
    for (const auto& item : program.items) {
        if (const auto* wf = std::get_if<ast::Workflow>(&item)) {
            WorkflowCompiler wc{guarantees, wf->name, {}, 0};
            contract.workflows[wf->name] = wc.compile(*wf);
        }
    }
    contract.guarantees = std::move(guarantees);
    contract.source = std::string(source);
    return contract;
}

}  // namespace complier
